src/backend/columnar/columnar_storage.c

```diff
@@ -0,0 +1,451 @@
+#define COLUMNAR_FIRST_ROW_NUMBER 0
+
+
+/*
+ * Map logical offsets to a physical page and offset where the data is kept.
+ */
+static inline SmgrAddr
+logical_to_smgr(uint64 logicalOffset)
+{
+	SmgrAddr addr;
+
+	addr.blockno = logicalOffset / COLUMNAR_BYTES_PER_PAGE;
+	addr.offset = SizeOfPageHeaderData + (logicalOffset % COLUMNAR_BYTES_PER_PAGE);
+
+	return addr;
+}
+
+
+/*
+ * Map a physical page and offset address to a logical address.
+ */
+static inline uint64
+smgr_to_logical(SmgrAddr addr)
+{
+	return COLUMNAR_BYTES_PER_PAGE * addr.blockno + addr.offset - SizeOfPageHeaderData;
+}
+
+
+static void ReadFromBlock(Relation rel, BlockNumber blockno, uint32 offset,
+						  char *buf, uint32 len);
+static void WriteToBlock(Relation rel, BlockNumber blockno, uint32 offset,
+						 char *buf, uint32 len, bool clear);
+static uint64 AlignReservation(uint64 prevReservation);
+
+
+/*
+ * ColumnarMetapageInit - initialize a new metapage in an empty relation
+ * with the given storageId.
+ *
+ * Caller must hold AccessExclusiveLock on the relation.
+ */
+void
+ColumnarMetapageInit(Relation rel, uint64 storageId)
+{
+	LockRelationForExtension(rel, ExclusiveLock);
+
+#ifdef USE_ASSERT_CHECKING
+	RelationOpenSmgr(rel);
+	uint64 nblocks = smgrnblocks(rel->rd_smgr, MAIN_FORKNUM);
+	Assert(nblocks == 0);
+#endif
+
+	/* create two pages */
+	Buffer newBuffer = ReadBuffer(rel, P_NEW);
+	ReleaseBuffer(newBuffer);
+	newBuffer = ReadBuffer(rel, P_NEW);
+	ReleaseBuffer(newBuffer);
+
+	ColumnarMetapage metapage = { 0 };
+	metapage.storageId = storageId;
+	metapage.versionMajor = COLUMNAR_VERSION_MAJOR;
+	metapage.versionMinor = COLUMNAR_VERSION_MINOR;
+	metapage.reservedStripeId = COLUMNAR_FIRST_STRIPE_ID;
+	metapage.reservedRowNumber = COLUMNAR_FIRST_ROW_NUMBER;
+	metapage.reservedOffset = ColumnarFirstLogicalOffset;
+
+	WriteToBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				 (char *) &metapage, sizeof(ColumnarMetapage),
+				 true /* clear because we are overwriting */);
+
+	UnlockRelationForExtension(rel, ExclusiveLock);
+}
+
+
+/*
+ * ColumnarMetapageRead - read the current contents of the metapage. No
+ * locks are held after return.
+ */
+void
+ColumnarMetapageRead(Relation rel, ColumnarMetapage *metapage)
+{
+	ReadFromBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				  (char *) metapage, sizeof(ColumnarMetapage));
+}
+
+
+/*
+ * ColumnarMetapageReserveData - reserve logical data offsets for writing.
+ */
+uint64
+ColumnarMetapageReserveData(Relation rel, uint64 amount)
+{
+	if (amount == 0)
+	{
+		return ColumnarInvalidLogicalOffset;
+	}
+
+	LockRelationForExtension(rel, ExclusiveLock);
+
+	ColumnarMetapage metapage;
+	ReadFromBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				  (char *) &metapage, sizeof(ColumnarMetapage));
+
+	uint64 alignedReservation = AlignReservation(metapage.reservedOffset);
+	uint64 nextReservation = alignedReservation + amount;
+	metapage.reservedOffset = nextReservation;
```

Comment:
what about aligning the `reservedOffset` beforehand (not when reserving the stripe but when writing `nextReservation` to metapage) ?

For example:
```suggestion
	uint64 currentReservation = metapage.reservedOffset;
	uint64 nextReservation = AlignOffset(currentReservation + amount);
	metapage.reservedOffset = nextReservation;
```

Reply:
The space is technically not reserved until someone asks for it. I'd prefer to just keep it the way it is unless there's a reason to change it.

Reply:
actually the main reason behind proposing that is to easily use `print_columnar_metapage` function in testing.
If we align the `reservedOffset` beforehand, then in regression tests we can check if the `print_columnar_metapage().nextOffset` matches `columnar.stripe` entry, as in below:

```sql
SELECT print_columnar_metapage(table);
-- metapage.reservedOffset is already aligned 

insert into table values(1);
-- verify if we use reservedOffset in next reservation
select file_offset = reservedOffset from columnar.stripe where table ..;
```

Reply:
I see, to check for gaps. I'm inclined to just remove the alignment completely in another PR, so that would make it moot.